Implement rocket with targeting system

Kuzenny-ns opened the pull request.

Kuzenny-ns committed:
Basic rocket behaviour

Kuzenny-ns committed:
Advanced targeting

Kuzenny-ns committed:
Merge conflict fix

Swap of scene files to resolve conflict

Kuzenny-ns committed:
Merge branch 'main' into kuzenny-ns/ledge-25-targeting-system

vdaysky commented:
⚠️**WARNING**⚠️ 
you became a victim of _passive-agressive Pull Request review process_ by **Me**. Get ready for tangible amout of nitpicks and naming conventions you never knew existed. If mental assistance is required, don't hesitate to call our suicide prevention hotline: +88005553535

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+    [SerializeField] private float minDistanceBoost;
+
+    [Header("Player Rocket Parameters")]
+    [SerializeField] private bool PlayerControlled;
```
Приватная переменная не должна начиняться с большой буквы

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+    [SerializeField] private GameObject targetSecondary;
+
+    //Some consts
+    private Vector3 inverseGravityVector = Vector3.up * 10f;
```
Константа, которая не константа, и могла бы быть локальной переменной. И вообще, чем `Physics.gravity` не нравится?

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+
+    private void OnCollisionEnter(Collision collision)
+    {
+        if(collision.gameObject.name == "Raptor")
```
Я не стал бы полагаться на имена. Кто-то переименует Raptor в Fighter и код поламался, никто не знает почему. Теги лучше, или может есть еще альтернативы

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+
+    void RocketVelocity()
+    {
+        // forces affecting the rocket shamelessly stolen from AirController
```
Украл однажды, укради и дважды, у нас новая система для физики

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+        rocketRigidbody.velocity = finalVelocity;
+    }
+
+    void RocketTargeing()
```
U Briish?

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+
+        //check if secondary target is seen
+        RaycastHit hit;
+        if (Physics.SphereCast(transform.position, 10f, transform.forward, out hit))
```
Inline var лучше

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+        RaycastHit hit;
+        if (Physics.SphereCast(transform.position, 10f, transform.forward, out hit))
+        {
+            if (hit.collider.name == targetSecondary.name + "(Clone)")
```
Опять имена

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+    }
+
+
+    private void RockeSteering()
```
U very briish?

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+
+    private void RockeSteering()
+    {
+        float M_x = Input.GetAxis("Mouse X") * SteeringSens * Time.deltaTime;
```
mX правильное именование

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+    void RocketTargeing()
+    {
+        float distanceBoost = 0f;
+        Vector3 target = targetMain.transform.position - transform.position;
```
Тут вектор как направление, ниже как позиция. Я понимаю, что для углов, но все равно почему и тут не взять позицию `targetMain.transform.position`

Kuzenny-ns replied:
Вектор всегда как направление, только при виде flares это направление задаётся позицией.

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,130 @@
+    // Update is called once per frame
+    void Update()
+    {
+        RocketVelocity();
```
Если использовать силы, то точно нужно перенести в FixedUpdate

vdaysky reviewed (changes requested):
Я бы отделил ракету от управления. Уж слишком по-разному работает ручное и автоматическое управление. В префабе я б оставил какой-то метод setTarget(Vector3 location) и всю математику, а потом из скрипта mainRocket или AirDefenceRocket я бы в каждом тике смотрел на цель (которая точно есть, в отличии от случая с ручным управлением) и проверял бы, выпуснала она flares или нет. Тогда и рейтрейсить не надо

Kuzenny-ns committed:
Requested changes

Kuzenny-ns committed:
Merge branch 'kuzenny-ns/ledge-25-targeting-system' of https://github.com/vdaysky/Leading-Edge into kuzenny-ns/ledge-25-targeting-system

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -10,121 +9,97 @@ public class RocketBehaviour : MonoBehaviour
         //check if secondary target is seen
-        RaycastHit hit;
-        if (Physics.SphereCast(transform.position, 10f, transform.forward, out hit))
+        if (Physics.SphereCast(transform.position, 10f, transform.forward, out RaycastHit hit))
```
Все равно не нравится мне это. Лучше как по мне это снести, и сделать метод
`setTarget(Vector3)`. Он будет вызываться из вне каждый тик. Математика останется вся та же, просто `target` приходит из вне. А потом внутри `RocketAiControl` ты будешь каждый тик делать `gameObejct.getComponent<AirController>.hasTriggeredFlares()` и если да, то `target += rocket.transform.up * randomUp + rocket.transform.left * randomLeft`, никакого трейса

Только сделай так, чтоб `randomUp` и `randomLeft` генерировались однажды за период работы flares, иначе нет смысла. И если вдруг `hasTriggeredFlares` перешло от `true` к `false`, то ты делаешь `setTarget(настоящая цель)`

`hasTriggeredFlares` для дебага можешь прописать максимально тупо, типа:

```csharp
public bool HasTriggeredFlares() {
    return _lastFlaresUsage + FlaresCooldownMs >= DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond
}
```

vdaysky commented on Assets/Scripts/RocketPlayerControl.cs:
```diff
@@ -0,0 +1,38 @@
+public class RocketPlayerControll : MonoBehaviour
+{
+    [Header("Player Rocket Parameters")]
+    [SerializeField] private bool playerControlled;
```
Хм, я чувствую ^C ^V

Kuzenny-ns replied:
Рефакторинг через ctrl+v и ctrl+c

vdaysky reviewed (changes requested):
Let Me Know если есть предложения получше или аргументы против

Kuzenny-ns committed:
Requested changes 2

Переименовал файл и поламало все изменения. Новый код на строчках 22, 31, 51, 64.

Kuzenny-ns committed:
Refactoring

Kuzenny-ns committed:
Relation change

vdaysky commented on Assets/Scripts/RocketAiControll.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class RocketAiControll : MonoBehaviour
```
Либо Controller либо Control

vdaysky commented on Assets/Scripts/RocketAiControll.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    [Header("Rocket Targets")]
+    [SerializeField] private GameObject targetMain;
+    [SerializeField] private GameObject targetSecondary;
```
Не используется

vdaysky commented on Assets/Scripts/RocketAiControll.cs:
```diff
@@ -0,0 +1,36 @@
+    void Update()
+    {
+        Vector3 target = targetMain.transform.position;
+        if (targetMain.GetComponent<AirController>().HasTriggeredFlares())
```
Дорого, лучше GetComponent однажды при старте

vdaysky commented on Assets/Scripts/RocketAiControll.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            createNewOffset = true;
+        }
+        transform.GetComponent<RocketBehaviour>().RotationToTarget(target);
```
аналогично

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,80 @@
+        //Probably most stupid math to clamp rotation
+        if (Vector3.Angle(transform.forward, targetDirection) > maxTurnAngle + distanceBoost)
+        {
+            Vector3 RotationVector = Vector3.Cross(transform.forward, targetDirection);
```
локальные переменные всегда с маленькой

vdaysky commented on Assets/Scripts/RocketAiControll.cs:
```diff
@@ -0,0 +1,36 @@
+    [SerializeField] private GameObject targetSecondary;
+
+    //flare offset
+    private float aimOffsetVertical = 0f;
```
приватные переменные должны начинаться с _

vdaysky commented on Assets/Scripts/RocketBehaviour.cs:
```diff
@@ -0,0 +1,80 @@
+        Vector3 fullLiftVector = -Physics.gravity * verticalSpeedCut;
+
+        // compute lift vector depending on engine and wing health
+        Vector3 liftVector = (fullLiftVector * (liftWithoutEngine * wingLiftFactor)) +
```
если упростить то всегда liftVector = fullLiftVector

vdaysky commented on Assets/Scripts/RocketPlayerControl.cs:
```diff
@@ -0,0 +1,20 @@
+        float mY = Input.GetAxis("Mouse Y") * steeringSens * Time.deltaTime;
+
+        Vector3 target = transform.position + transform.forward * 10 + transform.right * mX + transform.up * mY;
+        transform.GetComponent<RocketBehaviour>().RotationToTarget(target);
```
аналогично

Kuzenny-ns committed:
Requested changes 3

vdaysky reviewed (approved):
Nice rocket, awesome controllers
